### PR TITLE
Vagrant cloud provider not including ca.crt in secret

### LINKED ISSUE #10522: error: couldn't read version from server: Get https://10.245.1.2/api: dial tcp 10.245.1.2:443: connection refused
On latest Kubernetes version Release 0.20.1 Candidate I get the following when you setup cluster with "./cluster/kube-up.sh". This is on Mac OSX with VirtualBox, Vagrant

Any ideas, thanks

....
==> minion-3:  *  INFO: Running daemons_running()
==> minion-3:  *  INFO: Salt installed!
Wrote config for vagrant to /Users/user1/.kube/config
Each machine instance has been created/updated.
  Now waiting for the Salt provisioning process to complete on each machine.
  This can take some time based on your network, disk, and cpu speed.
  It is possible for an error to occur during Salt provision of cluster and this could loop forever.
Validating master
Validating minion-1
Validating minion-2
Validating minion-3
.....
Waiting for each minion to be registered with cloud provider
error: couldn't read version from server: Get https://10.245.1.2/api: dial tcp 10.245.1.2:443: connection refused

kubernetes user1$ ./cluster/kubectl.sh get nodes
error: couldn't read version from server: Get https://10.245.1.2/api: dial tcp 10.245.1.2:443: connection refused

### PR DESCRIPTION
**Pull request by derekwaynecarr**
This fixes the error in log when running kube2sky on vagrant setup where it was looking for a ca.crt file that was missing because the controller-manager was not including it in the default tokens for vagrant.

Fixes https://github.com/GoogleCloudPlatform/kubernetes/issues/10522

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
Still have to update cert: 
reflector.go:136] Failed to list *api.Endpoints: Get https://10.247.0.1:443/api/v1beta3/endpoints: x509: certificate is valid for 10.245.1.2, not 10.247.0.1

**Comment by roberthbailey** [comment]
@derekwaynecarr are you going to update the cert as part of this change?

**Comment by derekwaynecarr** [comment]
Yes.  I just ran out of time.

**Comment by zmerlynn** [comment]
Assigning to yourself. Assign someone else when this is ready.

**Comment by derekwaynecarr** [comment]
This is a targeted PR that affects just the Vagrant deployment's usage of Salt.

I know @eparis has a solution that would enable AWS, but my bias is that should be a follow-on if time is an issue (also at that time, we would need to bring back cert_ip in vagrant pillar).

Either way, this fixes kube2sky on vagrant.

**Comment by eparis** [comment]
My hopefully generic solution which solves the problem for all is at:
https://github.com/eparis/kubernetes/commit/ab52932ebeef385b429453d1fd012f8f97ada53e

But I think this patch is fine for 0.20.x   We should solve it for everyone, but this one only makes things a little worse. (I say worse, because the more we build scripts and hacks which are cloud provider only the more we build a brittle impossible to manage system, so I argue we do more in salt, less in 'provision', and less 'per cloud provider' everywhere.

**Comment by derekwaynecarr** [comment]
@zmerlynn - this is ready for your review... tried to make the most targeted fix to lower risk.  we can do a cleaner, more risky change as follow-on per @eparis

File: cluster/saltbase/salt/generate-cert/make-ca-cert.sh [hunk 1]
```diff
@@ -40,6 +42,16 @@ if [ "$cert_ip" == "_use_azure_dns_name_" ]; then
   use_cn=true
 fi
 
+if [ "$cert_ip" == "_use_vagrant_external_ip_" ]; then
+  cert_ip=$(/sbin/ifconfig eth1 | sed -n '2 p' | awk '{print $2}')
```
Comment (zmerlynn) [inline cluster/saltbase/salt/generate-cert/make-ca-cert.sh]: Always eth1?

Comment (zmerlynn) [inline cluster/saltbase/salt/generate-cert/make-ca-cert.sh]: Yeah, ignore me.

**Comment by zmerlynn** [comment]
LGTM

**Comment by zmerlynn** [comment]
cc @eparis for another review

**Comment by zmerlynn** [comment]
cc @dchen1107 for rubber stamp

**Comment by eparis** [comment]
LGTM

**Comment by derekwaynecarr** [comment]
Can this get merged today?

**Comment by zmerlynn** [comment]
@brendanburns: Since you seem to be awake, too, feel like signing off on this? I'd like to get it in ASAP, too.

**Comment by derekwaynecarr** [comment]
Friendly poke to merge?

**Comment by zmerlynn** [comment]
Rebase?

**Comment by yujuhong** [comment]
Pending a rebase and waiting on @dchen1107's "ok-to-merge"

**Comment by derekwaynecarr** [comment]
This is rebased and now works with @justinsb changes for how it uses the sans for certs.  Be good to see this merged ;-)

**Comment by zmerlynn** [comment]
Still LGTM

**Comment by eparis** [comment]
still LGTM

**Comment by zmerlynn** [comment]
This is hitting core in literally exactly one place and is otherwise confined to vagrant. Let's merge it.

**Comment by AlexeyKupershtokh** [comment]
I'm still getting
```
Validating master
Validating minion-1
........
Waiting for each minion to be registered with cloud provider
error: couldn't read version from server: Get https://10.245.1.2/api: x509: certificate is valid for 192.168.56.101, 10.247.0.1, not 10.245.1.2
```
when running
```
export KUBERNETES_PROVIDER=vagrant
cluster/kube-up.sh
```
on the v0.21.1.

**Comment by AlexeyKupershtokh** [comment]
@derekwaynecarr though I'm currently trying to repeat this on a clean installation.

**Comment by AlexeyKupershtokh** [comment]
Yep, the problem still persists.

**Comment by derekwaynecarr** [comment]
Are you using virtual box or VMware?

What host OS are you running from?

**Comment by derekwaynecarr** [comment]
@AlexeyKupershtokh - can you remove the ~/.kube directory and try to reproduce?

See my comment here for steps I followed:

https://github.com/GoogleCloudPlatform/kubernetes/issues/10900#issuecomment-120007639

**Comment by AlexeyKupershtokh** [comment]
I'm running ubuntu 14.04.2 lts 64 bit, virtualbox 4.3.28.
Trying to repeat using your way. I'll let you know any results in about 12h.

**Comment by derekwaynecarr** [comment]
Thanks for your help!  It's possible there is some race condition or
something that is not working deterministically all the time, so having
outside help with error details is always a big help.

On Thu, Jul 9, 2015 at 1:11 PM, Alexey Kupershtokh <notifications@github.com

**Comment by AlexeyKupershtokh** [comment]
Yep, still the same result.
I've also performed a test with logging the *SANS env variable.
It has the value of ```IP:10.247.0.1,DNS:kubernetes,DNS:kubernetes.default,DNS:kubernetes.default.svc,DNS:kubernetes.default.svc.cluster.local,DNS:kubernetes-master``` in my case.
As far as I understand the cert_ip is vagrant eth1 ip which is 192.168.56.101. 
So there is really no MASTER_IP (which is "10.245.1.2") that is used when accessing the cluster from the host machine.
If I add the MASTER_IP to the *SANS list, everything goes fine.

**Comment by derekwaynecarr** [comment]
So you are getting a different eth1 ip than me.  I will look to put the
certip grain back to resolve and cc you on the PR to verify it fixes for
you.

**Pull request merged**